test/integration/testlib/proc.py

```diff
@@ -0,0 +1,290 @@
+"""Classes for working with compiled instances of tinc and tincd binaries."""
+
+import os
+import time
+import typing as T
+import subprocess as subp
+from enum import Enum
+
+from . import check, path
+from .log import log
+from .script import TincScript, Script, ScriptType
+from .template import make_script, make_cmd_wrap
+from .util import random_string, random_port
+
+
+def _make_wd(name: str) -> str:
+    work_dir = os.path.join(path.test_wd, "data", name)
+    os.makedirs(work_dir, exist_ok=True)
+    return work_dir
+
+
+class Feature(Enum):
+    """Optional features supported by both tinc and tincd."""
+
+    COMP_LZ4 = "comp_lz4"
+    COMP_LZO = "comp_lzo"
+    COMP_ZLIB = "comp_zlib"
+    CURSES = "curses"
+    JUMBOGRAMS = "jumbograms"
+    LEGACY_PROTOCOL = "legacy_protocol"
+    LIBGCRYPT = "libgcrypt"
+    MINIUPNPC = "miniupnpc"
+    OPENSSL = "openssl"
+    READLINE = "readline"
+    TUNEMU = "tunemu"
+    UML = "uml"
+    VDE = "vde"
+
+
+class Tinc:
+    """Thin wrapper around Popen that simplifies running tinc/tincd
+    binaries by passing required arguments, checking exit codes, etc.
+    """
+
+    name: str
+    _work_dir: str
+    _port: T.Optional[int]
+    _scripts: T.Dict[str, TincScript]
+    _procs: T.List[subp.Popen]
+
+    def __init__(self, name: str = "") -> None:
+        self.name = name if name else random_string(10)
+        self._work_dir = _make_wd(self.name)
+        self._port = None
+        self._scripts = {}
+        self._procs = []
+
+    def randomize_port(self):
+        """Use a random port for this node."""
+        self._port = random_port()
+
+    def read_port(self) -> int:
+        """Read port used by tincd from its pidfile and update the _port field."""
+        pidfile = self.sub("pid")
+
+        for _ in range(30):
+            try:
+                log.debug("reading pidfile at %s", pidfile)
+                with open(pidfile, "r", encoding="utf-8") as f:
+                    content = f.read()
+
+                log.debug("found data %s", content)
+                _, _, _, token, port = content.split()
+                check.equals("port", token)
+
+                self._port = int(port)
+                return self._port
+            except OSError:
+                time.sleep(1)
```

Comment:
Why do we need to sleep and retry? As far as I can see, we only call `read_port()` after we already know the tinc daemon is up, so the pidfile should already be present.